### ISSUE TITLE
Login using credentials from IVLE

### ISSUE THREAD

**Issue by rrtheonlyone** [source-academy/cadet-frontend#68]
It would be good to have our services authenticated by IVLE. 

One way to go about doing this is to have the frontend call a URL from IVLE to get a session token. This token can then be stored in the state and used to verify whether a user is logged in.

**Comment by rrtheonlyone** [comment]
Worked on this (79997d9). Current progress:

1. Login Component has a button that calls IVLE with a redirect URL
     - To get it to work go into /mocks/IVLE_details.ts and add in your IVLE API Key obtained from  https://ivle.nus.edu.sg/LAPI/default.aspx
    - The callback URL is also in the same file /mocks/IVLE_details

2. Once there is a session token, an action will be dispatched and it will be stored in the state. 

Some changes I have done. 
* I have changed tslint to allow me to do console logging. 
* I moved history out to a seperate file so that I can do history.push('/device') for example

Some things to think about:
* The state gets erased when u refresh or go back to the same URL. The token does not persist?
* Should the session token be stored in the frontend?
* Currently there is a saga configured that does the history push. The saga can also call the backend to authenticate the user (whether he is taking CS1101S and his details). Question is should the backend handle all the IVLE calling or should it be done straight from the client-side?

@remo5000 @ningyuansg @evansb for your kind review and advice.

**Comment by ningyuansg** [comment]
> I have changed tslint to allow me to do console logging.

We prefer to disable console.logs in tslint. We use the react/redux dev tools for chrome, @remo5000 uses his IDE, and I work with `alert` instead (they pause execution, and are conveniently allowed in tslint). No issues with you using whatever you need to develop, of course, but preferably let's try not to modify tslint for that purpose. Would recommend you use something like git's [assume-unchanged](https://git-scm.com/docs/git-update-index#git-update-index---no-assume-unchanged) for that.

> I moved history out to a seperate file so that I can do history.push('/device') for example

Are there issues with importing into a `ts` file from a `tsx` file? I'm actually not very sure, but would it be sufficient to export the history object and import it where you need it? Nonetheless, I think `history.ts` (with the 'File' suffix) is a sufficiently descriptive name.

> The state gets erased when u refresh or go back to the same URL. The token does not persist?

Probably need cookies for that. They're not yet implemented, but they're definitely on the list. Basically, we haven't had a recent to use cookies so far (but persisting the auth token is a good one).

> Should the session token be stored in the frontend?

I don't think there's any way to avoid storing some form of token/key to identify the user, on the client. If you mean our API token to IVLE, I recall a meeting where @evansb said it's okay. If you _are_ talking about the user token, do see #53. Obfuscation of the session token would be a secondary feature IMO, since if a user shares their session token or tries to retrieve someone else's token from their browser, it's really on them (morally speaking)---kind of like sharing or trying to steal a password

Perhaps you could elaborate on what you mean?

> Currently there is a saga configured that does the history push. The saga can also call the backend to authenticate the user (whether he is taking CS1101S and his details). Question is should the backend handle all the IVLE calling or should it be done straight from the client-side?

Admittedly, I'm not very care on the details either. here's what I think happens.

1. User clicks login on frontend.
2. Frontend redirects user to IVLE login, together with our API token. Redirect is within the same tab.
3. Because IVLE recognises our API token, they render an appropriate login page with a callback url pointing to our frontend. User logins via this IVLE login page. IVLE redirects them via said callback url.
4. Frontend receives the callback url with some kind of IVLE identification token attached. Frontend sends the token to Backend.
5. Backend receives token. Fetches the relevant information (missions, status, etc.), gives it back to the Frontend.
6. Frontend receives this information, stores it in the redux store, renders the relevant information on the displayed component.

Again, not very certain on the details. Tagging @indocomsoft and @sreycodes  for confirmation.

**Comment by remo5000** [comment]
@rrtheonlyone I checked out the branch, and was able to alter the state using a url param, great to see that in place 👍 

A few queries:
- Is `actions/login` a placeholder?
- Why is pushing to the history necessary, if IVLE is going to cause a redirect anyways (causing an application refresh)?

**Comment by rrtheonlyone** [comment]
@ningyuansg and @remo5000  thanks for the comments. 

At @ningyuansg :
1. I have actually relooked the history part and realised I could just export it out straight from `index.ts` itself. So there is no need for the separate file! 
2. I have reversed the changes to tslint.
3. I dont think there is a difference between session token and user token. IVLE provides a session token once the user authenticates himself. Unless, there is another token? Yup I agree with your point about obfuscation and storing on client side.

At @remo5000:
1. Actions/Login was something I was experimenting with earlier. Removed it as well.
2. The push to history is necessary because once IVLE redirects, it will append the session token to the url. So it will be for example, `localhost:3000/device?token=1241515...` I want to remove the token and restore the url back to `localhost:3000/device`. 

I have updated my branch. Do let me know if there is anything else. 

I will look into improving this and integrating with the backend as well.
